qkchash/qkchash.cpp

```diff
@@ -94,11 +95,49 @@ void qkc_hash(
     }
 
     /*
-     * Truncate mix directly.
-     * Similar to ETH, compression and post-sha3 can be applied.
+     * Compress
+     */
+    for (uint32_t i = 0; i < result.size(); i++) {
+        result[i] = fnv64(mix[i * 2], mix[i * 2 + 1]);
+    }
+}
+
+void qkc_hash_sorted_list(
+        std::vector<uint64_t>& slist,
+        std::array<uint64_t, 8>& seed,
+        std::array<uint64_t, 8>& result) {
+    std::array<uint64_t, 16> mix;
+    for (uint32_t i = 0; i < mix.size(); i++) {
+        mix[i] = seed[i % seed.size()];
+    }
+
+    for (uint32_t i = 0; i < ACCESS_ROUND; i ++) {
+        std::array<uint64_t, 16> new_data;
+        uint64_t p = fnv64(i ^ seed[0], mix[i % mix.size()]);
+        for (uint32_t j = 0; j < mix.size(); j++) {
+            // Find the pth element and remove it
+            uint32_t idx = p % slist.size();
+            new_data[j] = slist[idx];
+            slist.erase(slist.begin() + idx);
+
+            // Generate random data and insert it
+            p = fnv64(p, new_data[j]);
+            slist.insert(std::upper_bound(slist.begin(), slist.end(), p), p);
```

Comment:
make sure we don't add duplicated element.

Reply:
Thanks for pointing this out.   Fixed in the new code.  Please take a look.

Reply:
I don't see the downside of keeping duplicates (except it would be different from the ordered set implementation)

Reply:
The downside is that for ordered set implementation, we have to keep the reference count of a key if a key is inserted twice or more.  In addition, implementing an ordered set with duplicated keys is more complicated.